### DISCUSSION TITLE
Scale Text inside NumericUpDown

### DISCUSSION THREAD

**Discussion by CaolanMullan** [AvaloniaUI/Avalonia#15057] [Q&A]
I have a NumericUpDown control inside a Grid which controls its size. However the text inside the control does not scale as the control gets smaller or bigger depending on the window size.

So far I have experimented with the viewbox control but this scales the entire NumericUpDown not just the text inside it resulting in over sized borders and up/down buttons.

Is there a way to only scale the text inside the NumericUpDown control?

Below is an example of what I have tired tried so far
``` .axaml
<Grid RowDefinitions="2*,3*,2*" ColumnDefinitions="*,10*,*">
      <Viewbox Grid.Row="1" Grid.Column="1">
           <NumericUpDown Grid.Row="1" Grid.Column="1" Value="10"/>
      </Viewbox>
</Grid>
```
Also any advice on scaling text based on space available/window size would be appreciated.

Thanks

**Comment by timunie** [comment]
FontSize with Converter should work

**Comment by CaolanMullan** [comment]
Thanks for the response @timunie 

I have read up on using Converters and experimented with applying custom logic to a bound value. However this still leave me with the problem of scaling up the font size inside the Converter.

I have no idea what is the most appropriate way to calculate the new font sized based on the space available. 

If you could point me in the right direction it would be appreciated.

**Reply by timunie** [reply]
the math is up to you.

**Reply by jp2masa** [reply]
Bootstrap's RFS is probably a good starting point: https://getbootstrap.com/docs/5.2/getting-started/rfs/.

**Comment by CaolanMullan** [comment]
I started by looking into how Maui scales text but Maui doesn't seem to have any built in support for text scaling, however with HTML/CSS I have had more luck,

CSS supports units of measure based on the "viewport" size where "One unit is 1% of one of the axes of the viewport." 

The four units are as follows:
- 1vw = 1% of viewport width
- 1vh = 1% of viewport height
- 1vmin = 1vw or 1vh, whichever is smaller
- 1vmax = 1vw or 1vh, whichever is larger

If supplied as a size for a font attribute each letter will take on a size corresponding to that percentage of the screen. So for example a viewport of width 'X' that directly contains text sized at 10vw (10% of the screen width) will fit 10 letters fit across the width of the screen. (There can be some give and take with non-mono spaced fonts).

I did some comparisons using a Blazor app scaling text vs Avalonia with me doing the math by hand for the font size. The output seem to line up pretty well so I'm going to experiment with how to implement this in code and I'll get back with the results.

sources:
https://css-tricks.com/viewport-sized-typography/
https://www.w3schools.com/css/css_font_size.asp

**Reply by timunie** [reply]
Then the converter would be easy imo. Give your ViewPort the name root. Then Bind to it like

{Binding #root.Bounds, Converter=..., ConverterParameter= 0.05}

**Reply by CaolanMullan** [reply]
Yes I was pretty surprised at how simple the math was as well. In retrospect I guess it make sense if you consider that a single character can be treated as 1 dimensional due to trying to maintain its aspect ratio. 

Our binding statements look more or less the same as well so I'll take that as a good sign. The cool thing is you can bind to the Width/Height of any element and treat it as the viewport which allows you scale text at different rates depending on the size of what they are inside off.

This method seems offers a lot more fine control/consistency than using "viewbox" in my experiments so far. I'm considering replacing all my viewboxes with it unless @timunie you can think of any reasons I shouldn't?

When I have more free time I'll post my implementations so anyone who wants to play with it/offer improvements can.

**Reply by timunie** [reply]
> This method seems offers a lot more fine control/consistency than using "viewbox" in my experiments so far. I'm considering replacing all my viewboxes with it unless @timunie you can think of any reasons I shouldn't?

Well depends on... If you want to to change FontSize only and keep BorderThickeness and other stuff as-is, use the converter method. In case you want everything to scale like in a "zooming", use ViewBox. For most situations, I prefer the FontSize approach but again this is more a design choice imo.
